packages/amplify_api_rest/android/src/test/kotlin/com/amazonaws/amplify/amplify_api_rest/AmplifyApiRestTest.kt

```diff
@@ -0,0 +1,414 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazonaws.amplify.amplify_datastore
+
+import com.amazonaws.amplify.amplify_api_rest.AmplifyApiRestPlugin
+import com.amplifyframework.api.ApiCategory
+import com.amplifyframework.api.ApiException
+import com.amplifyframework.api.aws.AWSApiPlugin
+import com.amplifyframework.api.rest.RestOperation
+import com.amplifyframework.api.rest.RestOptions
+import com.amplifyframework.api.rest.RestResponse
+
+import com.amplifyframework.core.Amplify
+import com.amplifyframework.core.Consumer
+import com.amplifyframework.hub.HubEvent
+import com.amplifyframework.hub.SubscriptionToken
+import io.flutter.plugin.common.MethodCall
+import io.flutter.plugin.common.MethodChannel
+import org.junit.Assert
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.ArgumentMatchers
+import org.mockito.Mockito
+import org.mockito.Mockito.*
+import org.mockito.invocation.Invocation
+import org.mockito.invocation.InvocationOnMock
+import org.robolectric.RobolectricTestRunner
+import org.robolectric.Shadows.shadowOf
+import java.lang.reflect.Field
+import java.lang.reflect.Modifier
+import java.util.concurrent.CountDownLatch
+
+@RunWith(RobolectricTestRunner::class)
+class AmplifyApiRestTest {
+    lateinit var flutterPlugin: AmplifyApiRestPlugin
+
+    private var mockApi = mock(ApiCategory::class.java)
+    private var mockAWSApiPlugin = mock(AWSApiPlugin::class.java)
+
+    private val mockResult: MethodChannel.Result = mock(MethodChannel.Result::class.java)
+    private val mockRestOperation: RestOperation = mock(RestOperation::class.java)
+
+    @Before
+    fun setup() {
+        flutterPlugin = AmplifyApiRestPlugin()
+
+        setFinalStatic(Amplify::class.java.getDeclaredField("API"), mockApi)
+    }
+
+    @Test
+    fun test_get_returns_success(){
+
+        var data = "{\"success\":\"get call succeed!\",\"url\":\"/items\"}".toByteArray()
+        var restResponse = RestResponse(200, data)
+
+        Mockito.doAnswer { invocation: InvocationOnMock ->
```

Comment:
`doAnswer` is useful to wire in mock behaviors when things get invoked. It isn't good practice to call a jUnit assertion inside of it. That's more like a Mockito `verify` call, where you're verifying that a call was made with expected arguments:
```java
ArgumentCaptor<String> stringCaptor = ArgumentCaptor.forClass(String.class);
verify(myObject.myFunction(stringCaptor.capture()))
assertEquals("The value I expected", stringCaptor.getValue());
```
^ Sorry it's in Java.

Also, minor, but you could write this first `doAnswer` line as:
```kotlin
doAnswer { invocation ->
```

Reply:
Hmm do you have a suggested alternative then?  Why isn't that a good practice?  In datastore we do the same exact thing so I assumed it was our testing standard.  In my tests and when developing the unit tests this approach seemed to work as well - I'm not too sure how else we could verify that the RestOptions object has been properly created from the map when the call was made. 

@jamesonwilliams

Reply:
Sure, try for something like this.

```kotlin
    @Test
    fun test_get_returns_success() {
        // Arrange some data.
        val data = JSONObject()
            .put("success", "get call succeeded!")
            .put("url", "items")
            .toString()
            .toByteArray()
            
        // When the get() API is invoked, fire a response on the callback
        // And return a mock operation.
        doAnswer {
            (invocation.arguments[1] as Consumer<RestResponse>)
                .accept(RestResponse(200, data))
            mockRestOperation
        }.`when`(mockApi)
            .get(
                any<RestOptions>(),
                any<Consumer<RestResponse>>(),
                any<Consumer<ApiException>>()
            )

        val arguments = mapOf(
            "restOptions" to mapOf("path" to "/items"),
            "code" to "someCode"
        )
        flutterPlugin.onMethodCall(MethodCall("get", arguments), mockResult)

        val optionsCaptor = ArgumentCaptor.forClass(RestOptions::class.java)
        verify(mockApi).get(
            optionsCaptor.capture(),
            any<Consumer<RestResponse>>(),
            any<Consumer<ApiException>>()
        )
        assertEquals(
            RestOptions.builder().addPath("/items").build(), 
            optionsCaptor.value
        )
        verify(mockResult)
            .success(mapOf("data" to restResponse.data.rawBytes))
    }
```

Alternately, instead of capturing the result in the new `verify` line, you could restrict the `doAnswer` match by changing `any<RestOptions>(),` to `eq(RestOptions.builder().addPath("/items").build())`.

Reply:
Thanks for that example code there, added that in.